composer/datasets/lm_dataset_hparams.py

```diff
@@ -90,93 +90,262 @@ def validate(self):
-            self.num_tokens = num_samples * tokens_per_sample
-        elif self.subsample_ratio == 1.0 and self.num_tokens == 0:
-            self.num_tokens = total_num_tokens
-        else:
-            log.warning('No subsampling going on!')
-
-        lm_datasets = lm_datasets.select(range(num_samples))  # type: ignore (thirdparty)
-        log.info(f'LM datasets: {lm_datasets}')
-        log.info(f'Subsample ratio: {self.subsample_ratio}')
-        log.info(f'Total number of samples: {num_samples:e}')
-        log.info(f'Total number of tokens: {self.num_tokens:e}')
-        dataset = lm_datasets
-
-        # for some tokenizers, e.g. GPT-2, they don't have padding tokens. Hence, we cannot use the LM collator.
-        if tokenizer.pad_token_id is None:
-            data_collator = transformers.default_data_collator
+            return build_synthetic_lm_dataloader(
+                synthetic_num_unique_samples=self.synthetic_num_unique_samples,
+                dataloader_hparams=dataloader_hparams,
+                tokenizer_name=self.tokenizer_name,  # type: ignore
+                batch_size=batch_size,
+                split=self.split,  # type: ignore
+                shuffle=self.shuffle,
+                drop_last=self.drop_last,
+                use_masked_lm=self.use_masked_lm,
+                num_tokens=self.num_tokens,
+                mlm_probability=self.mlm_probability,
+                subsample_ratio=self.subsample_ratio)
         else:
-            data_collator = transformers.DataCollatorForLanguageModeling(tokenizer=tokenizer,
-                                                                         mlm=self.use_masked_lm,
-                                                                         mlm_probability=self.mlm_probability)
-
-        sampler = dist.get_sampler(
-            cast(Dataset, dataset),  # HF datasets do not subclass torch datasets, so this cast is needed
-            drop_last=self.drop_last,
-            shuffle=self.shuffle)
-
-        return dataloader_hparams.initialize_object(
-            dataset=dataset,  # type: ignore
-            batch_size=batch_size,
-            sampler=sampler,
-            drop_last=self.drop_last,
-            collate_fn=data_collator)
+            return build_lm_dataloader(
+                datadir=self.datadir,
+                dataloader_hparams=dataloader_hparams,
+                tokenizer_name=self.tokenizer_name,  # type: ignore
+                batch_size=batch_size,
+                split=self.split,  # type: ignore
+                shuffle=self.shuffle,
+                drop_last=self.drop_last,
+                use_masked_lm=self.use_masked_lm,
+                num_tokens=self.num_tokens,
+                mlm_probability=self.mlm_probability,
+                subsample_ratio=self.subsample_ratio)
+
+
+def build_lm_dataloader(
+    datadir: List[str],
+    dataloader_hparams: DataLoaderHparams,
```

Comment:
@growlix can this be reworked to not use `DataloaderHparams`, but instead create the dataloader directly?

Reply:
Fix attempted. lmk if this is what you had in mind!